### PR TITLE
Nice transpiled output + tests

### PR DESCRIPTION
**Pull request by openorclose**
Makes it easier to parse the transpiled output.

### CODE REVIEW

File: src/__tests__/transpiled-code.ts [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+    // replace native[<number>] as they may be inconsistent
+    const replacedNative = transpiled.replace(/native\[\d+]/g, 'native')
+    // replace the line hiding globals as they may differ between environments
+    const replacedGlobalsLine = replacedNative.replace(/\n\(\(.*\)/, '\n(( <globals redacted> )')
+    // replace declaration of builtins since they're repetitive
+    const replacedBuiltins = replacedGlobalsLine.replace(
+      /\n *const \w+ = native\.builtins\.get\("\w+"\);/g,
+      ''
+    )
+    expect({ code, transpiled: replacedBuiltins }).toMatchSnapshot()
```
Comment (thomastanck) [inline src/__tests__/transpiled-code.ts]: Maybe better to return the promise and let the client code do the toMatchSnapshot(). Same amount of typing in both cases

File: src/__tests__/transpiled-code.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+  const context = mockContext(4)
+  const transpiled = transpile(parse(code, context)!, context.contextId).transpiled
+  // replace native[<number>] as they may be inconsistent
+  const replacedNative = transpiled.replace(/native\[\d+]/g, 'native')
+  // replace the line hiding globals as they may differ between environments
+  const replacedGlobalsLine = replacedNative.replace(/\n\(\(.*\)/, '\n(( <globals redacted> )')
+  expect({ code, transpiled: replacedGlobalsLine }).toMatchSnapshot()
+})
+
+test('Empty code returns empty string', expectTranspiledToMatchSnapshot(''))
```
Comment (thomastanck) [inline src/__tests__/transpiled-code.ts]: Then here you can have expectBlah(...).toEqual('')

(I'll pause here for now, will read the rest later)

File: src/__tests__/transpiled-code.ts (hunk 1 above)
Comment (openorclose) [reply src/__tests__/transpiled-code.ts]: What do you mean by this?

`return expect(..)` and then do `expectTranspiled('code').toMatchSnapshot()`?

File: src/__tests__/transpiled-code.ts (hunk 2 above)
Comment (openorclose) [reply src/__tests__/transpiled-code.ts]: won't this be then
```js
test('empty return empty', () => {
  expectBlah('').toBe('');
)
```
?

Comment (thomastanck) [reply src/__tests__/transpiled-code.ts]: Yeah, it's been a while since I touched jest :P

File: src/__tests__/transpiled-code.ts (hunk 1 above)
Comment (thomastanck) [reply src/__tests__/transpiled-code.ts]: Yep, exactly

Comment (thomastanck) [reply src/__tests__/transpiled-code.ts]: Looking back, I seem to have standardised `test*` which returns a promise that returns objects (which may or may not secretly check things with expect, e.g. status is successful), `snapshot*` which returns a similar promise but snapshots things first, and `expect*` which returns an expect object instantiated with one of the fields (and also may or may not secretly check other things with expect)

File: src/__tests__/__snapshots__/transpiled-code.ts.snap [hunk 3]
```diff
@@ -0,0 +1,514 @@
...
+Object {
+  "code": "function f() {
+  return null;
+}",
+  "transpiled": "const native = $$NATIVE_STORAGE;
+const boolOrErr = native.operators.get(\\"itselfIfBooleanElseError\\");
+const callIfFuncAndRightArgs = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
+const wrap = native.properTailCalls.wrap;
+(( <globals redacted> ) => {
+  return (() => {
```
Comment (thomastanck) [inline src/__tests__/__snapshots__/transpiled-code.ts.snap]: Now that I'm actually looking at the transpiled output, what's the purpose of this inner wrap?

File: src/__tests__/__snapshots__/transpiled-code.ts.snap [hunk 4]
```diff
@@ -0,0 +1,514 @@
...
+  "transpiled": "",
+}
+`;
+
+exports[`Ensure return values are of the form {isTail...} after function declarations are transformed 1`] = `
+Object {
+  "code": "function f() {
+  return null;
+}",
+  "transpiled": "const native = $$NATIVE_STORAGE;
```
Comment (thomastanck) [inline src/__tests__/__snapshots__/transpiled-code.ts.snap]: Actually I'm starting to think that these transpiled snippets are small enough that I can live with adding these to all the native-enabled tests... What do you think?

Comment (openorclose) [reply src/__tests__/__snapshots__/transpiled-code.ts.snap]: Yes this does sound reasonable.

File: src/__tests__/__snapshots__/transpiled-code.ts.snap (hunk 3 above)
Comment (openorclose) [reply src/__tests__/__snapshots__/transpiled-code.ts.snap]:
```js
const native = $$NATIVE_STORAGE;
const boolOrErr = native.operators.get(\\"itselfIfBooleanElseError\\");
const callIfFuncAndRightArgs = native.operators.get(\\"callIfFunctionAndRightArgumentsElseError\\");
const wrap = native.properTailCalls.wrap;
(( <globals redacted> ) => { // (A) list globals as parameters here to set to undefined at A2
  // if there's no inner wrap, 
  // const Number or const window would be illegal,
  // but it doesn't make sense to not allow them as variable names.
  return (() => {
    for (1; 1; 1) {}
    const lastStatementResult = eval(\\"1;\\");
    return lastStatementResult;
  })();
})(); // A2
```

Comment (thomastanck) [reply src/__tests__/__snapshots__/transpiled-code.ts.snap]: I see, thanks.

**Pull request merged**